File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/dns/DnsNMM.kt
package info.bagen.rust.plaoc.microService.sys.dns

import com.google.gson.JsonSyntaxException
import info.bagen.rust.plaoc.microService.core.*
import info.bagen.rust.plaoc.microService.helper.*
import info.bagen.rust.plaoc.microService.sys.jmm.JmmMetadata
import info.bagen.rust.plaoc.microService.sys.jmm.JsMicroModule
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.http4k.core.*
import org.http4k.lens.Query
import org.http4k.lens.string
import org.http4k.routing.bind
import org.http4k.routing.routes

inline fun debugDNS(tag: String, msg: Any = "", err: Throwable? = null) =
    printdebugln("fetch", tag, msg, err)

inline fun <K, V> MutableMap<K, V>.runExistOrPut(
    key: K, runExists:(V) -> Unit, defaultValue: () -> V
): V {
    val value = get(key)
    return if (value == null) {
        val answer = defaultValue()
        put(key, answer)
        answer
    } else {
        runExists(value)
        value
    }
}

class DnsNMM : NativeMicroModule("dns.sys.dweb") {
    private val installApps = mutableMapOf<Mmid, MicroModule>() // 已安装的应用
    private val runningApps = mutableMapOf<Mmid, MicroModule>() // 正在运行的应用

    suspend fun bootstrap() {
        bootstrapMicroModule(this)
    }

    /** 对等连接列表 */
    private val mmConnectsMap =
        mutableMapOf<MicroModule, MutableMap<Mmid, PromiseOut<ConnectResult>>>()
    private val mmConnectsMapLock = Mutex()

    /** 为两个mm建立 ipc 通讯 */
    private suspend fun connectTo(
        fromMM: MicroModule, toMmid: Mmid, reason: Request
    ) = mmConnectsMapLock.withLock {

        /** 一个互联实例表 */
        val connectsMap = mmConnectsMap.getOrPut(fromMM) { mutableMapOf() }

        /**
         * 一个互联实例
         */
        connectsMap.getOrPut(toMmid) {
            PromiseOut<ConnectResult>().also { po ->
                GlobalScope.launch(ioAsyncExceptionHandler) {
                    val toMM = open(toMmid);
                    debugFetch("DNS/connect", "${fromMM.mmid} => $toMmid")
                    val connects = connectMicroModules(fromMM, toMM, reason)
                    po.resolve(connects)
                    connects.ipcForFromMM.onClose {
                        mmConnectsMapLock.withLock {
                            connectsMap.remove(toMmid);
                        }
                    }
                }
            }
        }
    }.waitPromise()

    class MyDnsMicroModule(private val dnsMM: DnsNMM, private val fromMM: MicroModule) :
        DnsMicroModule {
        override fun install(mm: MicroModule) {
            // TODO 作用域保护
            dnsMM.install(mm)
        }

        override fun uninstall(mm: MicroModule) {
            // TODO 作用域保护
            dnsMM.uninstall(mm)
        }

        override suspend fun connect(
            mmid: Mmid, reason: Request?
        ): ConnectResult {
            // TODO 权限保护
            return dnsMM.connectTo(
                fromMM, mmid, reason ?: Request(Method.GET, Uri.of("file://$mmid"))
            )
        }

    }

    class MyBootstrapContext(override val dns: MyDnsMicroModule) : BootstrapContext {}

    suspend fun bootstrapMicroModule(fromMM: MicroModule) {
        GlobalScope.launch(ioAsyncExceptionHandler) {
            fromMM.bootstrap(MyBootstrapContext(MyDnsMicroModule(this@DnsNMM, fromMM)))
        }.join()
    }

    override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
        install(this)
        runningApps[this.mmid] = this

        /**
         * 对全局的自定义路由提供适配器
         * 对 nativeFetch 定义 file://xxx.dweb的解析
         */
        _afterShutdownSignal.listen(nativeFetchAdaptersManager.append { fromMM, request ->
            if (request.uri.scheme == "file" && request.uri.host.endsWith(".dweb")) {
                val mmid = request.uri.host
                debugFetch("DNS/fetchAdapter", "fromMM=${fromMM.mmid} >> requestMmid=$mmid: >> path=${request.uri.path} >> ${request.uri}")
                installApps[mmid]?.let {
                    val (fromIpc) = connectTo(fromMM, mmid, request)
                    return@let fromIpc.request(request)
                } ?: Response(Status.BAD_GATEWAY).body(request.uri.toString())
            } else null
        })


        val query_app_id = Query.string().required("app_id")
        val query_jmm_metadata = Query.string().required("jmmMetadata")

        /// 定义路由功能
        apiRouting = routes(
            // 打开应用
            "/open" bind Method.GET to defineHandler { request ->
                debugDNS("open/$mmid", request.uri.path)
                open(query_app_id(request))
                true
            },
            // 关闭应用
            // TODO 能否关闭一个应该应该由应用自己决定
            "/close" bind Method.GET to defineHandler { request ->
                debugDNS("close/$mmid", request.uri.path)
                close(query_app_id(request))
                true
            },
            // TODO 动态注册 JsMicroModule
            "/install" bind Method.GET to defineHandler { request, ipc ->
                debugDNS("install/${ipc.remote.mmid}", "query->${request.query("jmmMetadata")}")
                try {
                    gson.fromJson(query_jmm_metadata(request), JmmMetadata::class.java)
                } catch (e: JsonSyntaxException) {
                    debugDNS("install/${ipc.remote.mmid}", "fail -> ${e.message}")
                    null
                }?.also { jmmMetadata ->
                    install(JsMicroModule(jmmMetadata))
                }
                true
            })
        /// 启动 boot 模块
        GlobalScope.launch(ioAsyncExceptionHandler) {
            open("boot.sys.dweb")
        }
    }

    override suspend fun _shutdown() {
        installApps.forEach {
            it.value.shutdown()
        }
        installApps.clear()
    }

    /** 安装应用 */
    fun install(mm: MicroModule) {
        installApps[mm.mmid] = mm
    }

    /** 卸载应用 */
    fun uninstall(mm: MicroModule) {
        installApps.remove(mm.mmid)
    }

    /** 查询应用 */
    private suspend inline fun query(mmid: Mmid): MicroModule? {
        return installApps[mmid]
    }

    /** 打开应用 */
    suspend fun open(mmid: Mmid) : MicroModule {
        return runningApps.runExistOrPut(
            key = mmid,
            runExists = {
                if (it is JsMicroModule) {
                    it.nativeFetch(Uri.of("file://mwebview.sys.dweb/reOpen"))
                }
            }
        ) {
            query(mmid)?.also {
                bootstrapMicroModule(it)
            } ?: throw Exception("no found app: $mmid")
        }
    }

    /** 关闭应用 */
    suspend fun close(mmid: Mmid): Int {
        return runningApps.remove(mmid)?.let { microModule ->
            runCatching {

                mmConnectsMap[microModule]?.remove(mmid) // 将这个连接关闭
                microModule.shutdown()
                1
            }.getOrDefault(0)
        } ?: -1
    }
}

File: ios/BFExplorer/Plaoc/bfs/microModule/sys/dns/DnsNMM.swift
//
//  DwebDNS.swift
//  BFExplorer
//
//  Created by kingsword09 on 2023/1/30.
//

import Foundation
import Vapor

class DnsNMM: NativeMicroModule {
    private var mmMap: [Mmid:MicroModule] = [:]
    
    override init() {
        super.init()
        self.mmid = "dns.sys.dweb"
    }
    
    func bootstrap() async {
        await bootstrapMicroModule(fromMM: self)
    }
    
    /** 对等连接列表 */
    private var mmConnectsMap: [MicroModule:[Mmid:PromiseOut<ConnectResult>]] = [:]
//    private var mmConnectsMap: [MicroModule:Mmid] = [:]
    private let mmConnectMapLock = NSLock()
    /** 为两个mm建立 ipc 通讯 */
    private func connectTo(fromMM: MicroModule, toMmid: Mmid, reason: Request) async -> ConnectResult {
        await mmConnectMapLock.withLock {
            /** 一个互联实例表 */
            let connectsMap = mmConnectsMap[fromMM]
            
            /**
             * 一个互联实例
             */
            if connectsMap == nil {
                mmConnectsMap[fromMM] = [:]
            }
            
            let po = PromiseOut<ConnectResult>()
            Task {
                
                let toMM = await self.open(mmid: toMmid)
                print("DNS/connect", "\(fromMM.mmid) => \(toMmid)")
                let connects = await connectMicroModules(fromMM: fromMM, toMM: toMM, reason: reason)
                po.resolve(connects)
                _ = connects.ipcForFromMM.onClose {
                    _ = self.mmConnectMapLock.withLock {
                        self.mmConnectsMap[fromMM]!.removeValue(forKey: toMmid)
                    }
                    
                    return .OFF
                }
            }
            
            return po
        }.waitPromise()
    }
    
    class MyBootstrapContext: BootStrapContext {
        var dns: DnsMicroModule
        init(dns: MyDnsMicroModule) {
            self.dns = dns
        }
    }
    
    class MyDnsMicroModule: DnsMicroModule {
        private let dnsMM: DnsNMM
        private let fromMM: MicroModule
        
        init(dnsMM: DnsNMM, fromMM: MicroModule) {
            self.dnsMM = dnsMM
            self.fromMM = fromMM
        }
        
        func install(mm: MicroModule) {
            dnsMM.install(mm)
        }
        
        func uninstall(mm: MicroModule) {
            dnsMM.uninstall(mm)
        }
        
        func connect(mmid: Mmid, reason: Request?) async -> ConnectResult {
            return await dnsMM.connectTo(
                fromMM: fromMM,
                toMmid: mmid,
                reason: reason ?? Request.new(url: "file://\(mmid)"))
        }
    }
    
    func bootstrapMicroModule(fromMM: MicroModule) async {
//        Task {
//            await fromMM.bootstrap(bootstrapContext:MyBootstrapContext(dns: MyDnsMicroModule(dnsMM: self, fromMM: fromMM)))
//        }
        await fromMM.bootstrap(bootstrapContext:MyBootstrapContext(dns: MyDnsMicroModule(dnsMM: self, fromMM: fromMM)))
    }
    
    override func _bootstrap(bootstrapContext: BootStrapContext) async throws {
        install(self)
        running_apps[mmid] = self
        
        /**
         * 对全局的自定义路由提供适配器
         * 对 nativeFetch 定义 file://xxx.dweb的解析
         */
        _ = _afterShutdownSignal.listen(nativeFetchAdaptersManager.append { (fromMM, request) in
            if request.url.scheme == "file" && request.url.host != nil && request.url.host!.hasSuffix("dweb") {
                let mmid = request.url.host
                if mmid != nil {
                    let mm = self.mmMap[mmid!]
                    
                    if mm != nil {
                        let connectResult = await self.connectTo(fromMM: fromMM, toMmid: mmid!, reason: request)
                        
                        return await connectResult.ipcForFromMM.request(request: request)
                    }
                }
                
                return nil
            } else {
                return nil
            }
        })
        
        // 处理路由
        routerHandler()
        
        // 启动 boot 模块
        _ = await open(mmid: "boot.sys.dweb")
    }
    
    // 处理路由
    private func routerHandler() {
        // 保存路由处理方法，方便http和file协议请求调用
        let openRouteHandler: RouterHandler = { request, _ async in
            print("open/\(self.mmid)", request.url.path)
            guard let app_id = request.query[Mmid.self, at: "app_id"] else {
                return false
            }
            
            _ = await self.open(mmid: app_id)
            return true
        }
        let closeRouteHandler: RouterHandler = { request, _ async in
            guard let app_id = request.query[Mmid.self, at: "app_id"] else {
                return false
            }
            
            _ = await self.close(mmid: app_id)
            return true
        }
        apiRouting["\(self.mmid)/open"] = openRouteHandler
        apiRouting["\(self.mmid)/close"] = closeRouteHandler
        
        // 添加路由处理方法到http路由中
        let app = HttpServer.app
        let group = app.grouped("\(mmid)")
        let httpHandler: (Request) async throws -> Response = { request async in
            await self.defineHandler(request: request)
        }
        for pathComponent in ["open", "close"] {
            group.on(.GET, [PathComponent(stringLiteral: pathComponent)], use: httpHandler)
        }
    }
    
    override func _shutdown() async throws {
        for (_, mm) in mmMap {
            await mm.shutdown()
        }
        mmMap.removeAll()
    }
    
    private var running_apps: [Mmid:MicroModule] = [:]
    
    /** 安装应用 */
    func install(_ mm: MicroModule) {
        mmMap[mm.mmid] = mm
    }
    
    /** 卸载应用 */
    func uninstall(_ mm: MicroModule) {
        mmMap.removeValue(forKey: mm.mmid)
    }
    
    /** 查询应用 */
    private func query(mmid: Mmid) -> MicroModule? {
        return mmMap[mmid]
    }
    
    /** 打开应用 */
    private func open(mmid: Mmid) async -> MicroModule {
        var app = running_apps[mmid]
        if app != nil {
            return app!
        }
        
        app = query(mmid: mmid)
        
        if app == nil {
            fatalError("no found app: \(mmid)")
        }
        
        await bootstrapMicroModule(fromMM: app!)
        
        running_apps[mmid] = app!
        
        return app!
    }
    
    /** 关闭应用 */
    private func close(mmid: Mmid) async -> Int {
        let app = running_apps.removeValue(forKey: mmid)
        
        if app != nil {
            await app!.shutdown()
            return 1
        }
        
        return -1
    }
}
